File: kotlin/rotational-cipher/src/test/kotlin/RotationalCipherTest.kt
import org.junit.Test
import kotlin.test.assertEquals

class RotationalCipherTest {

    @Test
    fun testRotateLowercaseABy0() {
        val cipher = RotationalCipher(0)
        assertEquals("a", cipher.encode("a"))
    }

    @Test
    fun testRotateLowercaseABy1NoWrapAround() {
        val cipher = RotationalCipher(1)
        assertEquals("b", cipher.encode("a"))
    }

    @Test
    fun testRotateLowercaseABy26SingleWrapAround() {
        val cipher = RotationalCipher(26)
        assertEquals("a", cipher.encode("a"))
    }

    @Test
    fun testRotateLowercaseMBy13NoWrapAround() {
        val cipher = RotationalCipher(13)
        assertEquals("z", cipher.encode("m"))
    }

    @Test
    fun testRotateLowercaseNBy1SingleWrapAround() {
        val cipher = RotationalCipher(13)
        assertEquals("a", cipher.encode("n"))
    }

    @Test
    fun testRotateCapitalLettersNoWrapAround() {
        val cipher = RotationalCipher(5)
        assertEquals("TRL", cipher.encode("OMG"))
    }

    @Test
    fun testSpacesAreUnalteredByRotation() {
        val cipher = RotationalCipher(5)
        assertEquals("T R L", cipher.encode("O M G"))
    }

    @Test
    fun testNumbersAreUnalteredByRotation() {
        val cipher = RotationalCipher(4)
        assertEquals("Xiwxmrk 1 2 3 xiwxmrk", cipher.encode("Testing 1 2 3 testing"))
    }

    @Test
    fun testPunctuationIsUnalteredByRotation() {
        val cipher = RotationalCipher(21)
        assertEquals("Gzo'n zvo, Bmviyhv!", cipher.encode("Let's eat, Grandma!"))
    }

    @Test
    fun testAllLettersRotateCorrectly() {
        val cipher = RotationalCipher(13)
        assertEquals(
                "Gur dhvpx oebja sbk whzcf bire gur ynml qbt.",
                cipher.encode("The quick brown fox jumps over the lazy dog."))
    }

}

File: typescript/rotational-cipher/rotational-cipher.test.ts
import RotationalCipher from "./rotational-cipher"

describe("RotationalCipher", () => {
    it("rotate a by 1", () => {
        const result = RotationalCipher.rotate("a", 1)
        expect(result).toEqual("b")
    })

    it("rotate a by 26, same output as input", () => {
        const result = RotationalCipher.rotate("a", 26)
        expect(result).toEqual("a")
    })

    it("rotate a by 0, same output as input", () => {
        const result = RotationalCipher.rotate("a", 0)
        expect(result).toEqual("a")
    })

    it("rotate m by 13", () => {
        const result = RotationalCipher.rotate("m", 13)
        expect(result).toEqual("z")
    })

    it("rotate n by 13 with wrap around alphabet", () => {
        const result = RotationalCipher.rotate("n", 13)
        expect(result).toEqual("a")
    })

    it("rotate capital letters", () => {
        const result = RotationalCipher.rotate("OMG", 5)
        expect(result).toEqual("TRL")
    })

    it("rotate spaces", () => {
        const result = RotationalCipher.rotate("O M G", 5)
        expect(result).toEqual("T R L")
    })

    it("rotate numbers", () => {
        const result = RotationalCipher.rotate("Testing 1 2 3 testing", 4)
        expect(result).toEqual("Xiwxmrk 1 2 3 xiwxmrk")
    })

    it("rotate punctuation", () => {
        const result = RotationalCipher.rotate("Let's eat, Grandma!", 21)
        expect(result).toEqual("Gzo'n zvo, Bmviyhv!")
    })

    it("rotate all letters", () => {
        const result = RotationalCipher.rotate(
            "The quick brown fox jumps over the lazy dog.",
            13
        )
        expect(result).toEqual("Gur dhvpx oebja sbk whzcf bire gur ynml qbt.")
    })
})
